sql-plugin/src/main/scala/com/nvidia/spark/rapids/AvroDataReader.scala

```diff
@@ -59,22 +59,68 @@ private class SeekableInputStream(in: SeekableInput) extends InputStream with Se
 }
 
 /**
- * The header information of Avro file
+ * The header information of an Avro file.
  */
-class Header {
-  var meta = Map[String, Array[Byte]]()
-  var metaKeyList = ArrayBuffer[String]()
-  var sync = new Array[Byte](DataFileConstants.SYNC_SIZE)
-  var schema: Schema = _
-  private var firstBlockStart: Long = _
-
-  private[rapids] def update(schemaValue: String, firstBlockStart: Long) = {
-    schema = new Schema.Parser().setValidate(false).setValidateDefaults(false)
-      .parse(schemaValue)
-    this.firstBlockStart = firstBlockStart
+class Header private[rapids] {
+  private[rapids] val meta = mutable.Map[String, Array[Byte]]()
+  private[rapids] val sync = new Array[Byte](DataFileConstants.SYNC_SIZE)
+  private[rapids] var headerSize: Option[Long] = None
+
+  def firstBlockStart: Long = headerSize.getOrElse {
+    val out = new CountingOutputStream(NullOutputStream.NULL_OUTPUT_STREAM)
+    AvroFileWriter(out).writeHeader(this)
+    val newSize = out.getByteCount
+    headerSize = Some(newSize)
+    newSize
   }
 
-  def getFirstBlockStart: Long = firstBlockStart
+  @transient
+  lazy val schema: Schema = {
+    getMetaString(DataFileConstants.SCHEMA)
+      .map(s => new Schema.Parser().setValidateDefaults(false).setValidate(false).parse(s))
+      .orNull
+  }
+
+  private def getMetaString(key: String): Option[String] = {
+    meta.get(key).map(new String(_, StandardCharsets.UTF_8))
+  }
+}
+
+object Header {
+  /**
+   * Merge the metadata of the given headers.
+   * Note: It does not check the compatibility of the headers.
+   * @param headers whose metadata to be merged.
+   * @return the first header but having the new merged metadata, or
+   *         None if the input is empty.
+   */
+  def mergeMetadata(headers: Seq[Header]): Option[Header] = {
+    if (headers.isEmpty) {
+      None
+    } else if (headers.size == 1) {
+      Some(headers.head)
```

Comment:
The `reduce` should already produce what is desired without needing to handle it explicitly.
```suggestion
```